source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,319 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "absl/strings/str_cat.h"
+#include "fmt/printf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace PostgresProxy {
+
+/**
+ * Postgres messages are described in official Postgres documentation:
+ * https://www.postgresql.org/docs/12/protocol-message-formats.html
+ *
+ * Most of messages start with 1-byte message identifier followed by 4-bytes length field. Few
+ * messages are defined without starting 1-byte character and are used during well-defined initial
+ * stage of connection process.
+ *
+ * Messages are composed from various fields: 8, 16, 32-bit integers, String, Arrays, etc.
+ *
+ * Structures defined below have the same naming as types used in official Postgres documentation.
+ *
+ * Each structure has the following methods:
+ * read - to read number of bytes from received buffer. The number of bytes depends on structure
+ * type. toString - method returns displayable representation of the structure value.
+ *
+ */
+
+// Template for integer types.
+// Size of integer types is fixed and depends on the type of integer.
+template <typename T> class Int {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((data.length() - pos) < sizeof(T)) {
+      return false;
+    }
+    value_ = data.peekBEInt<T>(pos);
+    pos += sizeof(T);
+    left -= sizeof(T);
+    return true;
+  }
+
+  std::string toString() const { return fmt::format(getFormat(), value_); }
+
+  constexpr const char* getFormat() const { return "[{}]"; }
+  T get() const { return value_; }
+
+private:
+  T value_{};
+};
+
+using Int32 = Int<uint32_t>;
+using Int16 = Int<uint16_t>;
+using Int8 = Int<uint8_t>;
+
+// 8-bits character value.
+using Byte1 = Int<char>;
+
+// String type requires byte with zero value to indicate end of string.
+class String {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First find the terminating zero.
+    char zero = 0;
+    auto index = data.search(&zero, 1, pos);
+    if (index == -1) {
+      return false;
+    }
+
+    // Reserve that many bytes in the string.
+    auto size = index - pos;
```

Comment:
ditto. I'll give you a reason for this case, in addition to the style guide.

I'm always worried about unsigned arithmetic and overflow behavior  and I can't tell from context what the exact types are. So I can't simulate in my mind what the behavior will be.

Reply:
Agree. Corrected.